src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/InferParameterBindingInfoConvention.cs

```diff
@@ -132,12 +129,15 @@ internal void InferParameterModelPrefixes(ActionModel action)
         {
             foreach (var parameter in action.Parameters)
             {
+                // IEnumerable<IFormFile> parameters are collections of complex types and must be excluded. Otherwise
+                // FormFileModelBinder may attempt to bind file fields with an empty name (HTML disallows fields with
+                // an empty name). Users have historically applied [FromForm] to such parameters.
                 var bindingInfo = parameter.BindingInfo;
                 if (bindingInfo?.BindingSource != null &&
                     bindingInfo.BinderModelName == null &&
                     !bindingInfo.BindingSource.IsGreedy &&
                     !IsFormFile(parameter.ParameterType) &&
-                    IsComplexTypeParameter(parameter))
+                    !IsSimpleTypeOrSimpleCollection(parameter))
```

Comment:
I thought here's where we landed on as part of the offline discussion:

> I think the behavior we want is that collection typed parameters should by default be treated as [FromBody] for API controllers.

Reply:
Yes, exactly. In addition to that, we didn't want to revert the #7712 fix completely. This code here is about `BinderModelName` and not `BindingSource`.